### PR TITLE
[LWDM] LIVE-24193: zcash lib parse block shielded part

### PR DESCRIPTION
**Pull request by alevito**
### ✅ Checklist

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - the zcash-shielded library, and the changes in this PR, are not currently used anywhere in Ledger-Live. It will be enabled via feature flag for the purpose of development.

### 📝 Description

The method `ZCash.processBlock` adds the functionality of scanning a block for shielded transactions that are matching a given UFVK viewing key.
It returns a list of transactions for the given viewing key.

### ❓ Context

- **JIRA or GitHub link**: 

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 1]
```diff
@@ -18,6 +20,36 @@ export default class ZCash {
     return (endBlock - startBlock) * ZCash.AVERAGE_BLOCK_SYNC_TIME_MS;
   }
 
+  async processBlock(blockHex: string, viewingKey: string): Promise<string[]> {
```
Comment (alevito) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts]: rename to something like processShieldedBlock

Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: done in latest commit

Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: @may01 , the function `processBlock` has been renamed `findShieldedTxInBlock`

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    monitored: true;
+  };
+  valuePools: object[];
+  trees: { sapling: object; orchard: object };
+  previousblockhash: string;
+  nextblockhash: string;
+};
+
+export type RawTransaction = {
+  in_active_chain: boolean;
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:54]: Any reason for this to be in snake_case?

```suggestion
  inActiveChain: boolean;
```

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  };
+  valueBalance: number;
+  valueBalanceZat: number;
+  size: number;
+  time: number;
+  txid: string;
+  authdigest: string;
+  overwintered: true;
+  version: number;
+  versiongroupid: string;
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:106]: Same issue here. Does this need to be like that or can we keep to one convention like camelCase?

File: libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import { server } from "../mocks/node";
+import { HttpResponse, http } from "msw";
+
+beforeAll(() => server.listen({
+  onUnhandledRequest: 'error',
+}));
+afterEach(() => server.resetHandlers());
+afterAll(() => server.close());
+
+const JSON_RPC_SERVER = "http://localhost:18232";
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts]: Shouldn't this be pointing to the backend? It will fail in CI.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    chainValueZat: number;
+    monitored: true;
+  };
+  valuePools: object[];
+  trees: { sapling: object; orchard: object };
+  previousblockhash: string;
+  nextblockhash: string;
+};
+
+export type RawTransaction = {
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: Do we need all these props?

File: libs/coin-modules/zcash-shielded/mocks/handlers.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+              result: {
+                ...block,
+              },
+            });
+          }
+        }
+
+        return HttpResponse.json({
+          error: {
+            code: -5,
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/mocks/handlers.ts]: Lets replace error code with const.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 7]
```diff
@@ -18,6 +22,39 @@ export default class ZCash {
...
+    const transactions = block?.tx;
+    const decryptedTransactions = [];
+
+    // 3. retrieve each tx hash
+    if (transactions) {
+      for (const txId of transactions) {
+        const tx = await jsonRpcClient.getRawTransaction(txId);
+
+        // 4. call decryptTransaction for each tx hash containing orchard actions
+        if (tx?.orchard.actions.length) {
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts:80]: We should check for sapling as well.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 8]
```diff
@@ -18,6 +22,39 @@ export default class ZCash {
...
+          try {
+            const decryptedTx = await this.decryptTransaction(tx?.hex, viewingKey);
+            decryptedTransactions.push(toShieldedTransaction(tx, decryptedTx));
+          } catch (_e) {
+            console.warn("could not decrypt memo");
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts]: Will this be logged for each unsuccessful tx?

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
+import { JsonRpcClient } from "./jsonRpcClient";
+import { toShieldedTransaction } from "./shieldedTransaction";
 
 /**
  * ZCash API
  */
-
 export default class ZCash {
   static readonly AVERAGE_BLOCK_SYNC_TIME_MS = 5;
+  static readonly JSON_RPC_SERVER = "http://localhost:18232";
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts]: We should use currency config url here

File: libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts (hunk 4 above)
Comment (may01) [reply libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts]: does this points to the `libs/coin-modules/zcash-shielded/mocks/node.ts`?
Suggest to have only one variable that will set the expected server url
https://github.com/LedgerHQ/ledger-live/pull/13927/changes#diff-6e5c77ad08c5dd6654790d16c21346ec4448eb47f3a858e005cf1ac9d8a21eb5R4

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 5 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: These types have been derived from the HTTP responses served by the JSON RPC server. They are used in the JSON RPC client only.
We aren't currently using all the props in the response, but I thought it would be useful to have the full structure of the responses.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 2 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:54]: The reason why this is not in camel case is that this is exactly the response received from the server. The JsonRpcClient module is not restructuring the response retrieved from the server, but we could this functionality could be added.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 3 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:106]: See https://github.com/LedgerHQ/ledger-live/pull/13927/changes#r2728550296

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 7 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts:80]: I'm not 100% sure whether cipherscan's zcash-decoder can actually decrypt sapling transactions.
Are actions used also in sapling, or are they specific for orchard only?

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 8 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: Yes, it would. I can remove this and replace it with a unique log message emitted when the block is completely parsed.

File: libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts (hunk 4 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts]: The `JSON_RPC_SERVER` constant should be removed completely. The tests use a msw mock server which interceps requests to the backend and responds with mocked responses.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 7 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts:80]: This issue is now resolved.
We added support for sapling transactions in [LIVE-24171](https://github.com/LedgerHQ/ledger-live/pull/13887)

File: libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts (hunk 4 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts]: For the time being JSON_RPC_SERVER is defined in consts.
The `nodeUrl` string is currently passed as a parameter when the instance of ZCash is created by the consumer. Is this what we want?

File: libs/coin-modules/zcash-shielded/src/constants.ts [hunk 10]
```diff
@@ -0,0 +1,2 @@
+export const LOG_TYPE = "zcash-shielded";
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/constants.ts:1]: Can you rename this file to `constants.ts` to be consistent with other modules. I also created a `constants.ts` file in my PR and this way we can merge them.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: Are we going to use this type somewhere else? If so, it might make sense to move it into a `types.ts` file.

File: libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { DecryptedTransaction } from "@ledgerhq/zcash-decrypt";
+import { RawTransaction } from "./jsonRpcClient";
+
+export type ShieldedTransaction = {
```
Comment (semeano) [inline libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts:4]: Same here.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 13]
```diff
@@ -39,6 +49,48 @@ export default class ZCash {
...
   }
 
+  /**
+   * Scans a block for shielded transactions matching the given viewing key.
+   *
+   * @param {string} blockHex Block hex
+   * @param {string} viewingKey the UFVK - unified full viewing key.
+   * @returns {ShieldedTransaction[]} list of shielded transactions
+   */
+  async findShieldedTxsInBlock(
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts]: It will be better to pass block's data as an parameter. It will allow us to process batch of blocks in one take.

File: libs/coin-modules/zcash-shielded/src/constants.ts (hunk 10 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/constants.ts:1]: Thanks! Done in latest commit

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 11 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: @semeano 
Good idea, although I'd rather keep the type colocated with the context they belong to, in this case `jsonRpcClient.ts`, as they are specific of this module.
I'd rather re-export the types from `types.ts` in case they are ever needed by the consumer of the `zcash-shielded` module.

e.g.,
```typescrypt
// file: type.ts
export { RawTransaction } from "./src/jsonRpcClient";
```

Sounds good?

File: libs/coin-modules/zcash-shielded/src/constants.ts [hunk 14]
```diff
@@ -0,0 +1,2 @@
...
+export const JSON_RPC_SERVER = "http://localhost:18232";
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/constants.ts]: We should target proper ledger service.

File: libs/coin-modules/zcash-shielded/tests/mocks/handlers.ts
Comment (may01) [inline libs/coin-modules/zcash-shielded/tests/mocks/handlers.ts:1]: It's better to move mocks into `test` folder

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    });
+
+    if (data.error || !data.result) {
+      log(LOG_TYPE, `error: Empty dataonse result from server - ${data?.error?.message}`);
+    } else {
+      return data.result;
+    }
+  }
+
+  getBlock(blockHash: string) {
```
Comment (may01) [inline libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:140]: We also should be able to get range of blocks.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 11 above)
Comment (semeano) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: Sure!

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 15 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:140]: agreed: Might not be possible for the time being. Lets leave it aside for the time being.

File: libs/coin-modules/zcash-shielded/src/constants.ts (hunk 14 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/constants.ts]: Agreed: update the const value to the ledger service if it works. Otherwise leave localhost as example.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 13 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: Done: Updated findShieldedTxsInBlock to accept a Block argument, instead of a blockHash argument,.

File: libs/coin-modules/zcash-shielded/src/constants.ts (hunk 14 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/constants.ts]: Done: update the RPC server URL to ledger service.

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 15 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:140]: Verified that this is not currently possible.
ZCash 6.2.0 RPC getBlock call doesn't accept a block range  as argument.
See https://zcash.github.io/rpc/getblock.html

File: libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts (hunk 4 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/tests/jsonRcpClient.test.ts]: Resolved: Discussed over slack and agreed to keep `JSON_RPC_SERVER` in consts, and its value is the url to the legder node RPC service.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 9 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: Resolved: moved this to constants.ts

File: libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts (hunk 11 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/jsonRpcClient.ts:53]: Done in latest commit.
* `src/types.ts` exports all the module types.
* the compiled `lib/types.d.ts` is exposed by the package to the consumers - i.e., types field in package.json.

File: libs/coin-modules/zcash-shielded/mocks/handlers.ts (hunk 6 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/mocks/handlers.ts]: Done! Replaced with a const

File: libs/coin-modules/zcash-shielded/tests/mocks/handlers.ts
Comment (alevito) [reply libs/coin-modules/zcash-shielded/tests/mocks/handlers.ts:1]: done

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 8 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: this `try..catch` block was removed as the exceptions are already handled by `decryptTransaction`, which doesn't throw.

**Review by may01 — DISMISSED** [review]

File: libs/coin-modules/zcash-shielded/jest.config.js [hunk 16]
```diff
@@ -4,6 +4,7 @@ module.exports = {
   coverageDirectory: "coverage",
   testEnvironment: "node",
   testPathIgnorePatterns: ["lib/", "lib-es/", ".*\\.integration\\.test\\.[tj]s"],
+  transformIgnorePatterns: ["node_modules/(?!cipherscan)/"],
```
Comment (alevito) [inline libs/coin-modules/zcash-shielded/jest.config.js:7]: remove this transformIgnorePatterns as it's not required anymore

```suggestion
```

File: libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  id: string;
+  hex: string;
+  height: number;
+  blockHash: string;
+  time: number;
+  fee: number; // BigNumber as string
```
Comment (qperrot) [inline libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts]: I don't understand this comment

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 18]
```diff
@@ -1,5 +1,9 @@
-import { decrypt_tx, DecryptedTransaction } from "@ledgerhq/zcash-decrypt";
 import { log } from "@ledgerhq/logs";
+import { decrypt_tx, DecryptedTransaction } from "@ledgerhq/zcash-decrypt";
+import type { ShieldedTransaction } from "./shieldedTransaction";
+import { Block, JsonRpcClient } from "./jsonRpcClient";
+import { toShieldedTransaction } from "./shieldedTransaction";
```
Comment (qperrot) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts]: import from ./shieldedTransaction line 3 and 5

File: libs/coin-modules/zcash-shielded/src/ZCash.ts [hunk 19]
```diff
@@ -39,6 +49,45 @@ export default class ZCash {
...
+    const { block, viewingKey } = args;
+
+    // 1. get list of tx
+    const transactions = block?.tx;
+    const decryptedTransactions = [];
+
+    // 2. retrieve each tx hash
+    if (transactions) {
+      for (const txId of transactions) {
+        const tx = await this.jsonRpcClient.getRawTransaction(txId);
```
Comment (qperrot) [inline libs/coin-modules/zcash-shielded/src/ZCash.ts:77]: what if `txId` is "" ? is it possible ?

Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts:77]: Thanks, I think it's an edge case, but we need to handle it.
It should skip to the next iteration is `txId` is empty.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 18 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts]: Will fix this.

File: libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts (hunk 17 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/shieldedTransaction.ts]: Will fix this.

File: libs/coin-modules/zcash-shielded/src/ZCash.ts (hunk 19 above)
Comment (alevito) [reply libs/coin-modules/zcash-shielded/src/ZCash.ts:77]: I fixed this in commit 0fa3577116913e84e9e1b05e95cfe82c506231d7, and added a unit test for this edge case.
